### ISSUE TITLE
Android/Fragment Scopes with AndroidInjection.class

### ISSUE THREAD

**Issue by mariusc83** [google/dagger#640]
Hi,

I studied a bit the generated code following your guidelines for the AndroidInjectionModule and I could not see how we can keep using the Scope for our Activities/Fragments. 
The way I see you do not keep a strong reference to the AndroidInjector once you generate it when calling: AndroidInjection.inject(activity).

Thanks !!

**Comment by ronshapiro** [comment]
The scopes are effectively retained, since the `DispatchingAndroidInjector<Fragment>` is stored in `DaggerActivity`. The values of the `DAI<Fragment>` are subcomponents which retain the scope.

Give it a shot, and if you have any issues post back (or on StackOverflow)

**Comment by mariusc83** [comment]
Hmm,

I think I will have to give it another shot, last time I tried it and studied the generated code I did not see how this can work. For example for Activity, indeed you are keeping a reference of the DispatchingAndroidInjector inside the AppComponent but in order for a DoubleCheckProvider to work for my Activity the way I see it is that you will have to keep also a reference of the AndroidInjector once you build it: 
` AndroidInjector e = (AndroidInjector)Preconditions.checkNotNull(factory.create(instance), "%s.create(I) should not return null.", factory.getClass().getCanonicalName());
                e.inject(instance);`

I guess the trick is when binding my AndroidInjector.Factory:
` @Binds
   @IntoMap
   @ActivityKey(YourActivity.class)
   abstract AndroidInjector.Factory<? extends Activity>
       bindYourActivityInjectorFactory(YourActivitySubcomponent.Builder builder);`

to try to apply an @ActivityScope in order to use a DoubleCheck instead of a Provider.

Please let me know if I am wrong.

Tks !!

**Comment by mariusc83** [comment]
Thank you for your answer, I will try to use it more thoroughly and understand the way it works inside. I am rest assured now that you covered that part.

**Comment by ronshapiro** [comment]
There was a comment that got deleted which mentioned applying a scope to the `@Binds @IntoMap @ActivityKey` method. That scopes the _Builder_ which is almost definitely not what you want. We've added a check for this in ErrorProne to make sure people don't do it. Scoping the builder means the same instance of the builder will be kept around indefinitely, but you're asking for a built-component to reuse certain instances.

**Comment by mariusc83** [comment]
Actually this is the reason why I deleted the comment :). I just realized after posting it. But now I finally got it, all the "magic" happens in the other function inside the AndroidInjection.class:
`    public static void inject(Fragment fragment) {
        Preconditions.checkNotNull(fragment, "fragment");
        HasDispatchingFragmentInjector hasDispatchingFragmentInjector = findHasFragmentInjector(fragment);
        Log.d("dagger.android", String.format("An injector for %s was found in %s", new Object[]{fragment.getClass().getCanonicalName(), hasDispatchingFragmentInjector.getClass().getCanonicalName()}));
        DispatchingAndroidInjector fragmentInjector = hasDispatchingFragmentInjector.fragmentInjector();
        Preconditions.checkNotNull(fragmentInjector, "%s.fragmentInjector() returned null", hasDispatchingFragmentInjector.getClass().getCanonicalName());
        fragmentInjector.inject(fragment);
    }
`
Basically all the Fragments Subcomponents will be provided by the ActivityComponent and obviously that something is @ActivityScoped in there this will be provided as well inside the Fragments. 

Thanks !!
